### ISSUE TITLE
Important: Add "Global" and "Local" to the GUI

### ISSUE THREAD

**Issue by davidhedlund** [uBlock-LLC/uBlock#1371]
![global_local](https://cloud.githubusercontent.com/assets/6416073/7534861/1fd25918-f57f-11e4-8bab-f461f1c1ca78.png)

On the screenshot: The mouse cursor is in the second column. The "second column: global dynamic filtering rules, i.e. whatever rule appears in this column applies everywhere, on all sites:" - https://github.com/chrisaljoudi/uBlock/wiki/Dynamic-filtering:-quick-guide

Before I read that documentation I had no idea about this. Can you please add "Global" and "Local" to the GUI as I proposed on the modified screenshot @chrisaljoudi?

**Comment by atomGit** [comment]
i totally agree that the purpose of the columns needs to be specified on the UI, not just the documentation

rather than adding another row however, i wonder if the buttons/controls could display background text in addition to the plus/minus text?

perhaps a better idea might be to add a sort of status bar on the bottom of the UI that could hold all the text for the current tool-tips, plus the descrip. for the columns and the colors (green, grey, red)

**Comment by davidhedlund** [comment]
Numbers would be more precise than the current rating system that is based on + and -.

**Comment by Betsy25** [comment]
Yeah, probably for 85% of new try-outers that switch back to ABP, _the_ reason why they "don't understand" uBlock's way of working.

**Comment by gorhill** [comment]
> probably for 85% of new try-outers that switch back to ABP

90% of statistics are made up on the spot.

**Comment by ghost** [comment]
I agree with this. Before even installing the extension, I read the documentation and understood the use, but the two columns still creep me out and I have to think twice to remember that left=global, right=site.

**Comment by seanrand** [comment]
Not sure if I agree with the "important" part, but adding the information in an unobtrusive way wouldn't hurt. Although I personally would prefer it like this:
![screenshot-20150508t164636](https://cloud.githubusercontent.com/assets/929479/7538654/dc3c06f4-f5a1-11e4-961d-338465417649.png)
Edit: @ublockuser making all plus/minuses grey wasn't intentional, actually - I updated the screenshot. (and I didn't touch the font sizes, that's probably just a font quirk on my end)

**Comment by ghost** [comment]
@seanrand The smaller fonts for +/- and the moved color codes looks good. Nice idea.

**Comment by gorhill** [comment]
@seanrand Yes, best suggestion I've seen so far. It respects the low-noise level of extraneous information. The only issue will be to have translated strings fit in there as well. In my experience, the limited space is something which is overlooked sometimes by translators.

**Comment by 5t3f4n** [comment]
How about glyphs instead of text? Assures compatibility across all localizations.
A globe for global and a house or local network glyph for local or something like that.

Clarification could be added on hover if necessary, but at least translators have unlimited space that way.

**Comment by chrisaljoudi** [comment]
https://github.com/chrisaljoudi/uBlock/pull/1094

**Comment by ghost** [comment]
@chrisaljoudi I agree 100% with the use of icons with localized tooltips, and even your suggestion of moving the order of the columns. You most commonly want to unbreak ONE site, so you look at the blocked 3rd party host and unblock it for the current site. If the unblocking for the site is right next to the hostname, it makes a lot more sense than a spergy "global" column placed in the prime middle real estate, which is almost never used. Gorhill's comments (on that issue's page) about "users should rtfm to understand that left=global" is borderline autistic (as it doesn't care at all about newbies), and his claim that changing the order will confuse existing users goes against his rtfm claim; if power users rtfm and/or look at the icons they will understand the change. Even if they manage to miss both of those, they will still notice the change when they discover their newly added filters don't work as expected. So you have 100% support. Go for the more logical restructuring!

**Comment by davidhedlund** [comment]
@chrisaljoudi Hooray!

**Comment by ghost** [comment]
I agree with adding icons with tooltips, that's something that confused me when I first used the dynamic filtering pane of uBlock as well.

> You most commonly want to unbreak ONE site, so you look at the blocked 3rd party host and unblock it for the current site.



> [...] his claim that changing the order will confuse existing users goes against his rtfm claim; if power users rtfm and/or look at the icons they will understand the change.

I disagree. That's entirely subjective. One order is not objectively better than the other. If you change the order, everyone who already got used to the current order (and I'd risk saying that's most users, but don't take my word for it) would have to re-learn, so you haven't really gotten any further.

You could perhaps add a switch in the settings if people really want it, but I don't think it's that hard to learn.

> Gorhill's comments (on that issue's page) about "users should rtfm to understand that left=global" is borderline autistic (as it doesn't care at all about newbies) [...]

I suggest you to tone yourself down a bit. You just insulted all autistic people with that. However, I agree that telling people to RTFM for something simple like this is not very friendly at all.

**Comment by chrisaljoudi** [comment]
Regarding the swapping of the order:

<img src="https://imgs.xkcd.com/comics/workflow.png">

**Comment by chrisaljoudi** [comment]
Fixed in 3ca6daac73.

**Comment by gorhill** [comment]
> Regarding the swapping of the order:

Your point would make sense if at least your swapping of the order solved anything serious like "CPU overheating", but it doesn't.

**Comment by gorhill** [comment]
> However, I agree that telling people to RTFM for something simple like this is not very friendly at all.

The column headers won't save you to have to read the doc, there is much more to know about dynamic filtering than the column headers.

**Comment by chrisaljoudi** [comment]
@nyuszika7h @gorhill on a serious note, I understand your point completely. The goal is to determine a solution that optimizes for the overwhelming majority of users.

It is granted that:
- the amount of data we have on what users of dynamic filtering feel about this is not large
- it can be argued this sort of UI decision is somewhat "subjective" (that can be said about lots of UI decisions)

That said, I personally feel that the order makes a lot more sense once swapped. I, personally, haven't been able to intuitively internalize that the "middle" column is the global one while a more "outer" column is the site-specific "local" one.

I haven't heard much argument for keeping it other than the concern about users who might've already gotten used to it, but I have seen users saying they find the swapped order more intuitive.

All of the above is what justified the decision.

---

Thoughts:

The way the dynamic filtering pane expands immediately primes the user's perception to expect the hierarchy flow to be "right to left".

If the dynamic filtering pane expanded to the right (this is perhaps something to consider), the parameters might be different.

**Comment by ghost** [comment]
Rules are applied from top to bottom, from left to right. You should swap them for RTL languages only :)

**Comment by ghost** [comment]
@chrisaljoudi Again, you're making the assumption that everyone thinks the same way. I, for one, don't think about "middle" and "outer" at all; I think about "left" and "right", and for me, global being on the left makes more sense.

It doesn't make sense to flip the world upside down just because a few people complain. The fact that more people agree than disagree here (correct me if I'm wrong) means absolutely **nothing**, either.

**Comment by ghost** [comment]
@nyuszika7h: "It doesn't make sense to flip the world upside down". Strong words... Nobody did that. It was two columns swapping places, nothing Earth-shattering. And it was done in the name of helping out new users. All for a good cause.

@chrisaljoudi I support the change for this reason:
1. We are on site A.
2. We open the panel to see 3rd party hosts _used by site A_. (This point is _important_; everything in the matrix is related to the site _we are on_.)
3. Since we are on site A, we expect the blocking to relate to site A. It is the site we're on! Almost 100% of the time this isn't a "global" issue; it's a local-site one, such as manually unblocking a broken feature or blocking an annoyance.
4. The prime real estate of the firewall matrix is the middle column, which is immediately adjacent to the actual 3rd party host names. It's immediately visible next to the host names and requires the minimum amount of eye scanning and mouse movement.
5. We read left-to-right; left is the nearest and dearest (local), right is further away (global). So with this improved column order, we read from hostname -> local -> global. Very logical and intuitive. And with the new icons it's even more intuitive.

Excellent change.

**Comment by gorhill** [comment]
> We read left-to-right; left is the nearest and dearest (local), right is further away (global).

Perfect example of how subjective the whole thing is. I can as well argue that we write numbers with the most significant digit (broadest) at the left, and least significant at the right.

With regard to order, anything can be argued. I originally chose right for local because it is closest to the popup button and I assume this is the most used column. That simple.

**Comment by ghost** [comment]
@gorhill Well, then I can argue that you should read books from the last chapter first, right-to-left. ;-) Okay now it's getting confusing... (Oh and japanese people actually read that way, so they'd just say "of course we read the ending first!" ;-))

Anyway, about your last point: I don't understand what you mean by "the right column is the closest to the popup button"? Is that how it behaves on Chrome?

On Safari, I click the button and get this: http://i.imgur.com/V5693T3.png (perfectly centered under the button). And when I click the "requests blocked" header to expand the matrix, the whole popup window expands equally to the right and left (so that it remains centered): http://i.imgur.com/9e8Rku0.png

This means that after clicking the "requests blocked" header (which was in the center of the popup), my mouse cursor is immediately placed over the middle column (left/local, as we now call it). An easy way to confirm this is to look at the uBlock toolbar button in the 2nd screenshot, and scan downwards from there to see what the center of the window is; it's on the left column, and that's where the mouse is after expansion.

Does Chrome behave differently? Actually, the only reason I don't use Chrome on my Mac is because it's a lot more battery hungry and sluggish and lacks good gesture/zooming support on the trackpad. I would install it to check for myself if it wasn't so bad on Macs.

**Comment by gorhill** [comment]
> Is that how it behaves on Chrome?

That's what I get on Chromium and Firefox.

**Comment by ghost** [comment]
@gorhill So the popup window expands only leftwards on Chrome? Oh. I've learned something new today. That explains the original choice of putting the local blocking on the right, at least.

**Comment by ghost** [comment]
> Strong words... Nobody did that. It was two columns swapping places, nothing Earth-shattering.

It was a metaphor. I didn't mean it that way.

> And it was done in the name of helping out new users. All for a good cause.

As it has been explained numerous times, it's entirely subjective, so whether it's helping or slightly confusing (even with the icons) depends on who you ask.

**Comment by chrisaljoudi** [comment]
https://chrismatic.io/ublock/philosophy/

> ### 2. Development is considerate of feedback, but isn’t a democracy.
> 
> uBlock development pays a lot of attention to feedback and user stories, but isn’t a democracy. Having a single architectural vision will ensure a level of coherency and streamlining that wouldn’t otherwise be possible.

**Comment by davidhedlund** [comment]
@chrisaljoudi Can you please label this issue fixed?

**Comment by ghost** [comment]
@davidhedlund That's an entirely redundant label. There's wontfix and other labels if an issue is closed for another reason.

**Comment by davidhedlund** [comment]
@nyuszika7h Ok. When do you label and what do you label @chrisaljoudi?

Perhaps I can help better if I know that.
